Data/HashSet.hs

```diff
@@ -4,25 +4,68 @@
 #endif
 
 ------------------------------------------------------------------------
--- |
--- Module      :  Data.HashSet
--- Copyright   :  2011 Bryan O'Sullivan
--- License     :  BSD-style
--- Maintainer  :  johan.tibell@gmail.com
--- Stability   :  provisional
--- Portability :  portable
---
--- A set of /hashable/ values.  A set cannot contain duplicate items.
--- A 'HashSet' makes no guarantees as to the order of its elements.
---
--- The implementation is based on /hash array mapped trie/.  A
--- 'HashSet' is often faster than other tree-based set types,
--- especially when value comparison is expensive, as in the case of
--- strings.
---
--- Many operations have a average-case complexity of /O(log n)/.  The
--- implementation uses a large base (i.e. 16) so in practice these
--- operations are constant time.
+{-|
+Module      :  Data.HashSet
+Copyright   :  2011 Bryan O'Sullivan
+License     :  BSD-style
+Maintainer  :  johan.tibell@gmail.com
+Stability   :  provisional
+Portability :  portable
+
+= Introduction
+
+'HashSet' allows you to store /unique/ elements, providing efficient
+insertion, lookups, and deletion. If you are storing sets of @Int@s consider
+using the @Data.IntSet@ from the
+<https://hackage.haskell.org/packages/containers containers> package.
+A 'HashSet' makes no guarantees as to the order of its elements.
+
+== Immutability
+
+@HashSet@s are /immutable/ which means that any update functions do not modify
+the set you passed in, they create a new set. In order to keep the changes
+you need to assign it to a new variable.
+
+== Example
+
+@
+import qualified Data.HashSet as HashSet
+
+-- Create an empty HashSet.
+let s = HashSet.'HashSet.empty'
+
+-- Create a HashSet from a list of elements.
+let s1 = HashSet.'HashSet.fromList' ["a", "b"]
+
+-- Remove an entry from the HashSet @s1@, storing the new set in @s2@.
+let s2 = HashSet.'HashSet.delete' "a" s1
+
+-- Print the @HashSet@ @s1@, notice it still contains the deleted @"a"@.
+print s1
+> HashSet.'HashSet.fromList' ["a","b"]
+
+-- Print the modified @HashSet@, notice that it does not contain the deleted @"a"@.
+print s2
+> HashSet.'HashSet.fromList' ["b"]
+@
```

Comment:
See my commenthttps://github.com/haskell-unordered-containers/unordered-containers/pull/249#discussion_r436263795 on the other PR regarding the notation.

I think this would also be a good place to very briefly demonstrate some fundamental operations like `member`, `insert` and `delete`.

Reply:
Done, PTAL!